### PR TITLE
Flask: Create MinIO unlinked files endpoint

### LINKED ISSUE #155: Backend API to query MinIO for unlinked files that the current user is allowed to access
Task points: 5

May be blocked by #115

### PR DESCRIPTION
**Pull request by SamarthVP**
closes #155

### CODE REVIEW

File: flask/app/models.py [hunk 1]
```diff
@@ -9,26 +9,6 @@
...
-        self.password_hash = generate_password_hash(
-            password, method="pbkdf2:sha256:50000"
-        )
-
-    def check_password(self, password):
-        return check_password_hash(self.password_hash, password)
-
-    def get_id(self):
-        return self.user_id
-
```
Comment (kevinlul) [inline flask/app/models.py:31]: Could you move the `user_groups` table up instead of moving the `User` model down?

File: flask/app/buckets.py [hunk 2]
```diff
@@ -33,3 +39,46 @@ def get_object_info():
...
+        code = group.group_code.lower()
+        if code in all_bucket_names:
+            valid_bucket_names.append(code)
+
+    # Get all files in valid minio buckets
+    all_files = []
+    for bucket in valid_bucket_names:
+        objs = minioClient.list_objects(bucket)
+        for obj in objs:
+            all_files.append(obj.__dict__["object_name"])
```
Comment (kevinlul) [inline flask/app/buckets.py:70]: Does `obj.object_name` not work?

**Review by kevinlul — CHANGES_REQUESTED** [review]
How did you go about testing this endpoint?

File: flask/app/buckets.py (hunk 2 above)
Comment (SamarthVP) [reply flask/app/buckets.py]: Ah yes it does!

**Review by kevinlul — CHANGES_REQUESTED** [review]
The returned file names should be prefixed with `bucketname/` since object names are only unique within a bucket.

The runtime of this is `O(datasets * objects)` but if we use a dictionary (implemented as a hash table) to store either the known linked files or the unlinked files, then the runtime should be just `O(datasets + objects)` amortized due to constant search instead of linear search time. The same applies for finding valid bucket names but this isn't a big concern there since we don't expect to have many of those.

**Comment by SamarthVP** [comment]
Great idea, with those changes the run time is now effectively O(datasets + total objects) assuming the number of groups and buckets is small.

File: flask/app/buckets.py [hunk 3]
```diff
@@ -33,3 +39,46 @@ def get_object_info():
...
+    # Get all linked files
+    linked_files = {}
+    for dataset in models.Dataset.query.all():
+        if dataset.input_hpf_path is not None:
+            linked_files[dataset.input_hpf_path] = ":)"
+
+    # Put all unlinked files in new list
+    unlinked_files = []
+    for file_name in all_files:
+        if file_name[file_name.index('/')+1:] not in linked_files:
```
Comment (kevinlul) [inline flask/app/buckets.py:81]: This should just be `file_name` because we should be storing the combined bucket and object name that is unique

**Review by kevinlul — CHANGES_REQUESTED** [review]

File: flask/app/buckets.py (hunk 3 above)
Comment (SamarthVP) [reply flask/app/buckets.py:81]: oh so `input_hpf_path` will have the bucket prefix as well, got it

**Review by kevinlul — APPROVED** [review]
LGTM!

**Pull request merged**